documentation/interrupt_handling.md

```diff
@@ -109,15 +213,72 @@ When the monitor gets an FIQ exception it:
+8. Secure world issues an SMC to return to normal world
+9. Monitor saves secure world context and restores normal world context
+10. Does a return from exception into restored context
 
 ![FIQ received when SCR_NS is set](images/interrupt_handling/fiq.png "FIQ received when SCR_NS is set")
 
 ![FIQ received while processing an IRQ forwarded from secure world](images/interrupt_handling/irq_fiq.png "FIQ received while processing an IRQ forwarded from secure world")
 
-## Deliver FIQ to secure world when SCR_NS is cleared
+## 4.2. Deliver FIQ to secure world when SCR_NS is cleared
+
 Since SCR_FIQ is cleared when SCR_NS is cleared a FIQ will be delivered
 using the state vector (VBAR) in secure world. The FIQ is received as any
 other exception by Trusted OS, the monitor is not involved at all.
+
+# 5. Trusted Thread Scheduling
+
+## 5.1. Trusted Thread for Standard Services
+
+OP-TEE standard services are carried through standard SMC. Execution of these
+services can be interrupted by foreign interrupts. To suspend and restore the
+service execution, optee_os assignes a trusted thread at standard SMCs entry.
+
+The trusted thread
+terminates when optee_os returns to the normal world with a service completion
+status.
+
+A trusted thread execution can be interrupted by a native interrupt. In this case
+the native interrupt is handled by the interrupt exception handlers and once
+served, optee_os returns to the execution trusted thread.
+
+A trusted thread execution can be interrupted by a foreign interrupt. In this
+case, optee_os suspends the trusted thread and invokes the normal world through the
+Monitor (optee_os so-called RPC services). Resuming the trusted threads requires
+from the normal world to inovke optee_os with the expected RPC service status.
+
+A trusted thread execution can lead optee_os to invoke a service in normal
+world: access a file, get the REE current time, etc. The trusted thread is
+suspended/resumed during remote service execution.
+
+## 5.2. Scheduling Considerations
+
+When a trusted thread is interrupted by a foreign interrupt and when optee_os
+invokes a normal world service, the normal world gets the opportunity to
+reschedule the running applications. The trusted thread will resume only
+once the client application is scheduled back. Thus, a trusted thread execution
+follows the scheduling of the normal world caller context.
+
+Optee_os does not implement any thread scheduling. Each trusted thread is expected
+to track a service that is invoked from the normal world and should return to
+it with a execution status.
+
+The OP-TEE Linux driver (as implemented in drivers/tee/optee since Linux kernel
+4.12) is designed so that the Linux thread invoking OP-TEE gets
+assigned an trusted thread on TEE side. The execution of the trusted thread is
+tied to the execution of the caller Linux thread which is under the
+Linux kernel scheduling decision. Hence trusted threads are scheduled by the
```

Comment:
I guess, **hence** should not be used at the beginning of a sentence (at least in formal writing).